scripts/algorithms/knn.dml

```diff
@@ -0,0 +1,745 @@
+}
+print( "k value is " + k_value );
+# apply feature selection result
+if( select_feature == 1 ){
+    if( n_features > sum( m_selected_feature ) ){
+        Train = removeEmpty( target = Train,margin="cols",select = m_selected_feature);
+        Test = removeEmpty( target = Test,margin="cols",select = m_selected_feature );
+    }
+}
+
+NNR_matrix = naiveKNNsearch(Train,Test,k_value)
+write( NNR_matrix,NNRPath,format=FMT );
+
+# main part for knn predicted
+if( CLPath != " " ){
+    CL = read ( CLPath );
+    if(  CLTypePath == " " ){
+        stop( "Set Y but Miss Y_T for label type" );
+    }
+    CLType = read( CLTypePath );
+    i_cl_type = as.integer( as.scalar( CLType[1,1] ) );
+}
+CL_matrix = matrix( 0,nrow( Test ),1 );
+if( CLPath != " " ){
+    for( i in 1:nrow( NNR_matrix ) ){
+        NNR_tmp_matrix = matrix( 0,k_value,1 );
+        for( j in 1:k_value ){
+            NNR_tmp_matrix[j,1] = CL[as.scalar( NNR_matrix[i,j] ),1] ;
+        }
+        t_cl_value = 0.0 #test data label value
+        if( i_cl_type == 2 ){
+            t_cl_value = as.scalar( rowIndexMax( t( table( NNR_tmp_matrix,1 ) ) ) )
+            #if( PRPath == " " ){
+            #    print( "cate is : " + t_cl_value );
+            #}
+        }else{
+            if ( predict_con_tg == 0){
+                t_cl_value = mean( NNR_tmp_matrix );
+            } else if(predict_con_tg == 1){
+                t_cl_value = median( NNR_tmp_matrix );
+            }
+            #if( PRPath == " " ){
+            #    print( "value is : " + t_cl_value );
+            #}
+        }
+        CL_matrix[i,1] = t_cl_value;
+    }
+}
+if( PRPath != " " ){
+    write( CL_matrix,PRPath,format=FMT );
+}
+
+#naive knn search implement
+naiveKNNsearch = function(
+    matrix[double] P,
+    matrix[double] Q,
+    integer K
+)return(
+    matrix[double] O
+){
+    num_records = nrow (P);
+    num_features = ncol (P);
+    num_queries = nrow (Q);
+    Qt = t(Q);
+    PQt = P %*% Qt;
+    P2 = rowSums (P ^ 2);
+    D = -2 * PQt + P2;
+    if (K == 1) {
+        Dt = t(D);
+        O = rowIndexMin (Dt);
+    } else {
+        O = matrix (0, rows = num_queries, cols = K);
+        parfor (i in 1:num_queries) {
+            D_sorted=order(target=D[,i], by=1, decreasing=FALSE, index.return=TRUE);
+            O[i,] = t(D_sorted[1:K,1]);
+        }   
+    }   
+}
+
+# prepareKNNData:
+#   Do data prepare - dummy for category variable and [-1,1] transform for continues variable 
+# Argument:
+# in_m_data                     input matrix as features
+# in_m_data_type                input matrix with 1 row denoting whether columns in X are 
+#                               continuous ( =1 ) or categorical ( =2 )
+# in_i_trans_continues          Option flag for continuous feature transformed:
+#                               0 = do not transform continuous variable;
+#                               1 = transform continuous variable;
+prepareKNNData = function ( 
```

Comment:
The current data preparation can be  expensive specially for categorical varibaes. Would it be possible to handle categorical variables directly without dummy coding? In this case we ca decide within different methods how to handle categorical variables, e.g. in distance computation and so on.